File: ex15advclasses.m
% ex15: Advanced Classes, MATLAB

classdef ex15advclasses < handle & ex6classes
    properties (Access = protected)
        baz  % Create protected property
    end

    methods
        function obj = ex15advclasses(foo, bar)
            obj = obj@ex6classes(foo, bar);  % Call super constructor
            obj.examine(foo, bar);
            obj.baz = length(foo);
        end

        function check(obj, baz)
            disp('Checking...');
            check@ex6classes(obj, baz);  % Call super method
        end
    end

    methods (Access = protected)
        function empty = examine(obj)
            empty = isempty(obj.foo) || isempty(obj.bar);
        end
    end

    methods (Static)
        function test()
            example = ex15advclasses([252 253], [105 106]);
            gcd = example.process();
            example.check(gcd);
        end
    end
end

File: ex15advclasses.py
# ex15: Advanced Classes, Python

from ex6classes import Example6


class Example15(object, Example6):

    def __init__(self, foo, bar):
        super().__init__(foo, bar)  # Call super constructor
        self._examine(foo, bar)
        self._baz = 3.14
        self._examine()  # Create protected attribute

    def check(self, baz):
        print('Checking...')
        super().check(baz)  # Call super method

    def _examine(self):
        return (not self.foo) or (not self.bar)

    @staticmethod
    def test():
        example = Example15([252, 253], [105, 106])
        gcd = example.process()
        example.check(gcd)
        return example


if __name__ == "__main__":
    ex1 = Example15.test()
    ex1.qux = 3.14  # Create a new attribute
